File: Journey/Level_0_fundamentals/matrix_ops.py
'''
Docstring for Journey.Level_0_fundamentals.matrix_ops

Task: Implement a matrix class with multiplication and transpose.

* matrix multiplication
* transpose
* scalar multiply

'''

r, c = map(int, input().split())
c2, p = map(int, input().split())

if c != c2:
    raise ValueError("Invalid matrix dimensions")

mat1 = []
for _ in range(r):
    mat1.append(list(map(float, input().split())))

mat2 = []
for _ in range(c):
    mat2.append(list(map(float, input().split())))

s = float(input())

mat_mul = [[0.0] * p for _ in range(r)]
for i in range(r):
    for k in range(c):
        aik = mat1[i][k]
        for j in range(p):
            mat_mul[i][j] += aik * mat2[k][j]

transpose1 = []
for j in range(c):
    row = []
    for i in range(r):
        row.append(mat1[i][j])
    transpose1.append(row)

scalar1 = []
for i in range(r):
    row = []
    for j in range(c):
        row.append(mat1[i][j] * s)
    scalar1.append(row)

for row in mat_mul:
    print(*row)

for row in transpose1:
    print(*row)

for row in scalar1:
    print(*row)

File: Journey/Level_0_fundamentals/matrix_ops.cpp
#include <iostream>
using namespace std;

int main() {
    int r, c, c2, p;
    cin >> r >> c;
    cin >> c2 >> p;

    if (c != c2) {
        cout << "Invalid matrix dimensions";
        return 0;
    }

    double mat1[r][c], mat2[c][p];

    for (int i = 0; i < r; i++)
        for (int j = 0; j < c; j++)
            cin >> mat1[i][j];

    for (int i = 0; i < c; i++)
        for (int j = 0; j < p; j++)
            cin >> mat2[i][j];

    double result[r][p];
    for (int i = 0; i < r; i++)
        for (int j = 0; j < p; j++)
            result[i][j] = 0;

    for (int i = 0; i < r; i++)
        for (int k = 0; k < c; k++)
            for (int j = 0; j < p; j++)
                result[i][j] += mat1[i][k] * mat2[k][j];

    for (int i = 0; i < r; i++) {
        for (int j = 0; j < p; j++)
            cout << result[i][j] << " ";
        cout << endl;
    }

    for (int j = 0; j < c; j++) {
        for (int i = 0; i < r; i++)
            cout << mat1[i][j] << " ";
        cout << endl;
    }

    double s;
    cin >> s;

    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++)
            cout << mat1[i][j] * s << " ";
        cout << endl;
    }

    return 0;
}
